[Team-22][iOS] 1주차-2: 컬렉션 뷰 적용, 리팩토링

junu0516 opened the pull request:
안녕하세요! team-22 주방,Jed 입니다.

이번 작업에서는 아래의 내용을 적용해봤습니다.
- [x] JSONHandler의 메소드를 static 아닌 인스턴스메소드로 사용하도록 수정
- [x] 컬렉션뷰 적용 및 커스텀 셀 구현
- [x] 컬렉션뷰에 모델에서 미리 파싱해놓은 json 데이터 연결해서 출력 확인 

이후 진행 예정인 작업 내용은 아래와 같습니다.
- Alamofire을 사용한 네트워크 요청 및 GET요청 후 응답받은 이미지 데이터 적용
- 메인 화면 내 이미 구현된 컬렉션 뷰 화면 지속적으로 다듬기 및 미구현된 뷰 구현
- 네트워크 통신을 담당할 NetworkHandler 구현

**** 
- 지난번에 안티 싱글톤에 대해 말씀해주셔서 학습 후 아래와 같은 이유로 싱글톤 사용을 하지 않기로 결정했습니다.
    - 추후 NetworkHandler, JSONHandler의 구현을 완성한 후 이를 위한 테스트 메소드를 작성해야 하는데 싱글톤 패턴을 사용할 경우에는 mock object를 만들기가 어려움
    - 싱글톤 객체가 '상태'를 가져야 하는 상황에서, 전역범위로 객체를 공유하면 SOLID에 위배됨
- 우선은 JSONHandler를 사용할 때 인스턴스를 만들어 사용하도록 수정했고, 현재는 Ordering(모델) 내부에 구체타입으로 선언되있지만 곧 이를 추상타입으로 바꿔 적용할 예정입니다.(제네릭 관련해서 문법 오류가 일부 발생해서 현재 이에 대해 살펴보고 있습니다.)

junu0516 committed:
Chore: 디렉토리 구조 변경, 불필요한 주석 제거

junu0516 committed:
Refactor: Category 열거형의 rawValue 사용 제거

junu0516 committed:
Chore: 프로젝트 파일 업데이트 반영

junu0516 committed:
[#3] Refactor: Response 구조체에 제네릭 추가

- 객체배열인 body 안에는 다양한 Codable을 채택한 구체타입이 들어갈 수 있기 때문에 제네릭을 사용하도록 수정

junu0516 committed:
Fix: 알 수 없는 오류나서 디렉토리 구조 재수정

junu0516 committed:
Chore: ViewController -> MainViewController 이름 변경

junu0516 committed:
[#4] Refactor: foodList 제거 후 foodMap으로 사용하도록 수정

- [String:Food] 딕셔너리가 아닌 [Category:[String:Food]] 타입으로 사용

junu0516 committed:
[#4] Test: OrderingTest Ordering 변화된 로직에 맞춰 수정

junu0516 committed:
[#4] Refactor: Food 구조체 생성자 추가

- 디코딩 컨네이너 관련 생성자를 추가했기 때문에, 다른 생성자도 다시 명시해야 함

junu0516 committed:
[#4] Test: FoodTest 변화된 Food생성자에 맞춰 수정

LeeJoobang committed:
[#4] Feat: CollectionView 속성 추가

- Navigation Controller 연결

LeeJoobang committed:
[#4] Feat: FoodCollectionView 삽입 및 출력

LeeJoobang committed:
[#4] Feat: FoodCollectionView 내 FoodList 내역 콘솔 출력

junu0516 committed:
[#4] Refactor: Ordering 내부 foodMap 속성 private으로 변경

LeeJoobang committed:
[#4] Feat: FoodCollectionViewCell 내 UIImageView, UILabel 추가

LeeJoobang committed:
[#4] Feat: Cell 내  Food 정보 표시

junu0516 committed:
[#4] Refactor: JSONHandler의 static 메소드를 모두 인스턴스 메소드로 수정

- 싱글톤 사용을 하지 않기 위해 인스턴스 만들어 쓰기로 결정
- 추후 프로토콜을 만들어서 추상타입 인스턴스를 사용하도록 변경해야 함

junu0516 committed:
[#4] Refactor: MainViewController 내부의 viewDidLoad() 로직 별도 메소드로 분기 처리

ITzombietux commented on SideDish/SideDish/Commmon/JSONHandler.swift:
```diff
@@ -1,15 +1,16 @@
+    private let encoder = JSONEncoder()
+    private let decoder = JSONDecoder()
     
-    static func convertJSONToObject<T: Decodable>(from data: Data, to targetType: T.Type)-> T?{
```
JSONHandler 개체의 행동인 convertJSONToObject / converObjectToJSON 메서드를 제공하는 Protocol을 사용하는 방법에 대해서 알아보시면 좀 더 원하시는 tastable한 코드가 될거 같아요!!

junu0516 replied:
넵!! 현재 __`JSONHandlable`__ 프로토콜을 정의해서 모델에서 추상타입으로 핸들러를 가지도록 수정하려고 작업 진행중입니다.

ITzombietux commented on SideDish/SideDish/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,65 @@
+
+class MainViewController: UIViewController {
+    
+    private let ordering = Ordering()
```
Ordering과 같은 개체 역시 선언과 동시에 초기화하는 방식보다는 init()을 활용한다면 좀 더 유연한 코드가 될거 같아요! 추가적으로 위에서 드린 코멘트의 연장선인데 Ordering개체를 직접적으로 참조하는 것이 아니라 Ordering의 역할에 대한 Protocol을 만들어서 Protocol을 ViewController에서 속성으로 갖는다면 훨씬 더 좋을거 같아요!

junu0516 replied:
넵!! 구체타입보다는 추상타입으로 속성을 가지는 식으로 개선진행하겠습니다

ITzombietux commented on SideDish/SideDish/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,65 @@
+    }
+}
+
+extension MainViewController: UICollectionViewDelegate, UICollectionViewDataSource{
```
UICollectionViewDelegate, UICollectionViewDataSource와 같은 프로토콜을 MainViewController가 대리자 역할을 해주기 보다는 CustomDelegate / CustomDatasource를 만든다면 MainViewControllor가 아니라 다른 뷰컨트롤러에서도 재사용할 수 있을 것이고 한층 더 추상화된 코드를 작성할 수 있겠죠! "Custom UICollectionViewDataSource (with generic)" 같은 키워드를 학습해보시는 것을 추천해요!

junu0516 replied:
Custom 클래스에서 Datasource를 관리하는 건 생각하지 못했습니다! 알려주신 키워드로 검색해서, CollectionView 관련 프로토콜 채택에 대한 책임을 별도로 가지는 계층을 구성해보려 합니다! 감사합니다 :-)

ITzombietux commented on SideDish/SideDish/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,65 @@
+
+extension MainViewController: UICollectionViewDelegate, UICollectionViewDataSource{
+    func numberOfSections(in collectionView: UICollectionView) -> Int {
+        return Category.allCases.count
```
이렇게 직접적으로 Category count를 가져오는 것은 좋지 않을 수 있어요. Category 개체를 직접적으로 참조하지 않는 방법에 대해서도 protocol같은 키워드를 사용하여 알아보시면 좋을거 같아요! Ordering같은 개체에서 Category 개체에 물어보거나 할 수도 있겠죠!

junu0516 replied:
넵 우선은 Category 열거형에 직접 접근하는 것을 막고자 Ordering의 함수를 호출해서 카테고리 수에 해당하는 값을 리턴받는 식으로 수정했습니다.
뭔가 컨트롤러가 모델에 값을 요청해서 받아오는 것이 자연스러운 것 같아서 이런식으로 했고, 여기에 프로토콜을 어떻게 적용할 수 있을 지에 대해서 추후 더 고민해봐야할 것 같습니다!

ITzombietux commented on SideDish/SideDish/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,65 @@
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        guard let cell = foodCollectionView .dequeueReusableCell(withReuseIdentifier: "FoodCollectionViewCell", for: indexPath) as? FoodCollectionViewCell else { return UICollectionViewCell() }
+        cell.addViews()
+        cell.setLayout()
```
cell addViews / setLayout 메서드를 이 곳에서 호출하는 이유를 알 수 있을까요~?

LeeJoobang replied:
collectionView(메소드) 안에서 cell을 그린다고 생각해서 여기서 호출(addViews - foodImageView, FoodInforeamtionStackView, FoodNameLabel, foodDescriptionLabel, foodNormalPriceLabel, foodSpecialPriceLabel 자체를 Cell의 ContentView에 표시한다는 것 / setLayout - 오토레이아웃을 통해 foodImageView 위치,  foodInformationStackView 위치 설정) 하는 것이 맞다고 생각했습니다.

ITzombietux commented on SideDish/SideDish/Model/Ordering.swift:
```diff
@@ -1,47 +1,54 @@
-    var foodMap: [Category:[Food]] = [:]
-    var foodList: [String:Food] = [:]
+    private var foodMap: [Category:[String:Food]] = [:]
     private (set) var selectedMenu: Food?
```
Ordering같은 개체를 porotocol을 채택한 개체로 만든다면  private(set) 로 만들기 보단 메서드를 통하여 selectedMenu을 제공할 수 있을거 같네요!

junu0516 replied:
넵 해당 메소드는 디테일 뷰를 구현하면서 아마 본격적으로 쓰게 될 듯한데, 그 때 ordering을 보고 완전히 숨길 수 있는 부분에 대해서 private으로 일괄 처리하고, ordering의 동작들을 추상화한 형태의 프로토콜을 정의하려 합니다!

우선 private(set)으로 한 부분은 오로지 getter만이 필요한 속성에 대해서만 임시로 처리했습니다.

ITzombietux reviewed (commented):
프로젝트를 진행할 수 있는 시간이 하루인데 대단히 수고많으셨습니다.! 추가로 커밋을 하신다고 하여 종합적인 피드백은 추후에 드릴게요! 전반적으로 개체가 개체를 직접적으로 참조하고 있는 부분을 Protocol을 사용하여 개선한다면 굉장히 좋울거 같아요!

junu0516 committed:
[#4] Chore: MainViewController에서 분기된 개별 함수 모두 호출

junu0516 committed:
[#8] Feat: JSONHandlable 프로토콜 구현

junu0516 committed:
[#8] Refactor: Ordering에서 jsonHandler 속성이 추상타입이 되도록 수정

- JSONHandlable 타입으로 선언한 후 JSONHandler 인스턴스를 할당

junu0516 committed:
[#8] Add: RepositoryApplicable, RepositoryDelegate 프로토콜 추가

- 백그라운드 데이터 및 JSON파싱을 담당하는 레이어 및 델리게이트의 추상타입을 정의

junu0516 committed:
[#8] Add: Category 열거형을 별도의 파일로 분리해서 추가

- Ordering.swift에 있던 열거형 선언은 제거

junu0516 committed:
[#8] Chore: model 디렉토리명 변경하면서 기존 디렉토리에 있던 파일들 제거

- 파일들이 변경된 디렉토리명 하위로 이동된 것

junu0516 committed:
[#8] Chore: 변경된 디렉토리로 Response.swift 이동

junu0516 committed:
[#8] Feat: RepositoryApplicable을 채택한 Repository 구현 및 변경된 디렉토리로 이동

- 현재는 프로젝트 폴더 안에 있는 json 파일을 읽어들인 후 이를 디코딩하는 역할을 수행하는 별도의 레이어로 기능

junu0516 committed:
[#8] Refactor: Ordering 내부의 JSON 파싱 관련 로직 모두 제거, Repository 추상타입 속성 추가

junu0516 committed:
[#8] Chore: Food, Ordering 변경된 디렉토리 하위로 이동

junu0516 committed:
[#4] Chore: MainViewController 주석 제거, 불필요한 함수 호출 제거, Category 직접 접근하는 부분 제거

- ordering.getSampleFoodList()는 ordering 내부에서 자체적으로 호출하므로 뷰컨트롤러에서는 해당 함수의 호출을 제거
- 카테고리 수를 구할 때는 Category에 뷰컨트롤러가 직접 접근하지 않고, ordering으로부터 받아오도록 흐름을 수정

junu0516 committed:
[#8] Refacor: Testable한 구조를 위해 Ordering 내에서 repository 속성 초기화하는 로직 변경, MainViewController에서 Ordering 주입하는 방식 변경

- 생성자에서 사용할 repository 구체타입을 받아서, 샘플 데이터를 업데이트 하는 방식으로 수정
- 이후 Ordering 테스트 코드 내에서는 Repository가 아닌 미리 짜여진 MockRepository를 받아서 사용할 수 있는 것
- MainViewController의 ordering 속성은 변화된 Ordering의 생성자에 맞춰서, viewDidLoad 시점에 초기화되도록 수정

junu0516 committed:
[#8] Feat: MockRepository 추가 및 구현

- 내부에 임의의 Food타입 객체를 하나 가지고, 이를 사용하는 가상의 메소드를 하나만 구현

junu0516 committed:
[#8] Test: JSONHandlerTest 추가, OrderingTest 수정

- JSONHandlerTest는 임의의 Food타입 객체를 가지고 사용하도록 테스트함수 작성
- OrderingTest는 기존과 달리 구현한 MockRepository를 주입받는 Ordering객체를 사용해서 사용하도록 테스트함수를 수정

junu0516 commented:
> 프로젝트를 진행할 수 있는 시간이 하루인데 대단히 수고많으셨습니다.! 추가로 커밋을 하신다고 하여 종합적인 피드백은 추후에 드릴게요! 전반적으로 개체가 개체를 직접적으로 참조하고 있는 부분을 Protocol을 사용하여 개선한다면 굉장히 좋울거 같아요!

안녕하세요! 바쁘신 와중에 추가 커밋까지 받아주셔서 정말 감사합니다 (_ _)
현재 기존에 슬랙 메시지 통해 말한 대로 코드 형식 및 디렉토리 구조 등을 변경하는 부분을 반영했고,
기존에 주신 피드백 중 __`객체가 다른 객체를 추상타입으로 참조`__ 하라고 하신 부분과 __`JSON을 핸들링하는 객체의 함수를 테스트 가능하게 사용`__ 하라고 하신 부분에 대해 추가로 작업해서 커밋했습니다!
(커밋량이 좀 많아진 것 같아서 이후의 모델 및 뷰 관련 작업은 다음 PR때 올리도록 하겠습니다! )

추가로 기존에 남겨주신 리뷰 중에 뷰컨트롤러나 컬렉션 뷰 관련하여 질문하신 부분은 주방께서 작성하신 코드라 작성한 코드 흐름 한 번 다시 보고 답변 달 것 같습니다!
